Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
         for c in 0..<column {
           if columnsWithDiagnostics.contains(c) {
-            preMessage.append("│")
+            message.append("│")
           } else {
-            preMessage.append(" ")
+            message.append(" ")
           }
         }
+        return message
+      }()
 
-        for diag in diags.dropLast(1) {
-          annotatedSource.append("\(preMessage)├─ \(diagnosticDecorator.decorateDiagnosticMessage(diag.diagMessage))\n")
-        }
-        annotatedSource.append("\(preMessage)╰─ \(diagnosticDecorator.decorateDiagnosticMessage(diags.last!.diagMessage))\n")
+      func createAnnotatedSourceLine(with diagnostic: FlattenDiagnostic, using marker: String) -> String {
+        let colorizedMessage = diagnosticDecorator.decorateMessage(diagnostic.message, basedOnSeverity: diagnostic.severity)
+        return diagnosticMessagePrefix + marker + " " + colorizedMessage + "\n"
       }
 
-      // Add suffix text.
-      annotatedSource.append(annotatedLine.suffixText)
-      if annotatedSource.last != "\n" {
-        annotatedSource.append("\n")
+      // Append all but the last diagnostic with the appropriate marker
+      for diagnostic in diagnostics.dropLast() {
+        let annotation = createAnnotatedSourceLine(with: diagnostic, using: "├─")
+        annotatedSource.append(annotation)
       }
+
+      // Append the last diagnostic with a different marker; use force unwrapping because it always exists
+      let lastDiagnostic = diagnostics.last!
+      let finalAnnotation = createAnnotatedSourceLine(with: lastDiagnostic, using: "╰─")
+      annotatedSource.append(finalAnnotation)
+    }
+  }
+}
+
+extension DiagnosticsFormatter {
+  /// Generates a string containing source code with annotated diagnostics.
+  ///
+  /// - SeeAlso: ``annotatedSource(inSyntaxTree:withDiagnostics:usingIndentString:appendingSuffixTexts:employingSourceLocationConverter:)``
+  static func annotatedSource(
+    inSyntaxTree tree: some SyntaxProtocol,
+    withDiagnostics diagnostics: [Diagnostic],
+    usingIndentation indentString: String = "",
+    appendingSuffixTexts suffixTexts: [AbsolutePosition: String] = [:],
+    usingSourceLocationConverter sourceLocationConverter: SourceLocationConverter? = nil,
+    limitedToContextRange contextRange: ContextRange = .limited(2),
+    withDiagnosticDecorator decorator: DiagnosticDecorator = BasicDiagnosticDecorator()
+  ) -> String {
+    let formatter = Self(contextRange: contextRange, diagnosticDecorator: decorator)
+    return formatter.annotatedSource(
+      inSyntaxTree: tree,
+      withDiagnostics: diagnostics,
+      usingIndentString: indentString,
+      appendingSuffixTexts: suffixTexts,
+      employingSourceLocationConverter: sourceLocationConverter
+    )
+  }
+}
+
+/// Extends `DiagnosticsFormatter` with internal helper types for working with diagnostics.
+extension DiagnosticsFormatter {
```

Comment:
Since all of these are `private` and thus not visible outside of this file, I think we can also make them top-level types, reducing the nesting level by one.

Reply:
Sure, why not; done.